kaiax/builder/bundle.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2024 The Kaia Authors
+// This file is part of the Kaia library.
+//
+// The Kaia library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Kaia library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package builder
+
+import (
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/common"
+)
+
+type Bundle struct {
+	// each element can be either *types.Transaction, or TxGenerator
+	BundleTxs []interface{}
+
+	// BundleTxs is placed AFTER the target tx. If empty hash, it is placed at the very front.
+	TargetTxHash common.Hash
+}
+
+// Has checks if the bundle contains a tx with the given hash.
+func (b *Bundle) Has(hash common.Hash) bool {
+	for _, txOrGen := range b.BundleTxs {
+		switch tx := txOrGen.(type) {
+		case *types.Transaction:
+			if tx.Hash() == hash {
+				return true
+			}
+		default:
+			continue
+		}
+	}
+
+	return false
+}
+
+// IsConflict checks if newBundle conflicts with current bundle.
+func (b *Bundle) IsConflict(newBundle *Bundle) bool {
+	// 1. Check for same target tx hash
+	if b.TargetTxHash == newBundle.TargetTxHash {
+		return true
+	}
+
```

Comment:
Though unlikely, check for empty BundleTxs
```
// 2-1. Empty bundleTxs does not conflict with other transactions
if len(b.BundleTxs) == 0 {
  return false
}
```

Reply:
added cb60037